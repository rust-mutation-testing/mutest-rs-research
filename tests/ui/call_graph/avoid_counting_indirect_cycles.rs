//@ print-call-graph
//@ print-targets
//@ stdout
//@ stderr: empty

fn recursive_function(i: usize) {
    if i == 1 { return; }
    make_indirect_recursive_call(i - 1);
}

fn non_recursive_function_a() {}
fn non_recursive_function_b() {}

fn non_recursive_function_c() {
    non_recursive_function_a();
    non_recursive_function_b();
}

fn make_indirect_recursive_call(i: usize) {
    recursive_function(i);

    non_recursive_function_a();
    non_recursive_function_b();
    non_recursive_function_c();
}

#[test]
fn test() {
    recursive_function(3);
}
