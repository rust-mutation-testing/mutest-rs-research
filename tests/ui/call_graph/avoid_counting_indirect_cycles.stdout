
@@@ call graph @@@

entry points:

test test
  -> recursive_function at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:6:1: 6:32 (#0)

nested calls at distance 1:

recursive_function at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:6:1: 6:32 (#0)
  -> make_indirect_recursive_call at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:19:1: 19:42 (#0)

nested calls at distance 2:

make_indirect_recursive_call at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:19:1: 19:42 (#0)
  -> recursive_function at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:6:1: 6:32 (#0)
  -> non_recursive_function_a at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:11:1: 11:30 (#0)
  -> non_recursive_function_b at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:12:1: 12:30 (#0)
  -> non_recursive_function_c at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:14:1: 14:30 (#0)

nested calls at distance 3:

non_recursive_function_c at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:14:1: 14:30 (#0)
  -> non_recursive_function_a at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:11:1: 11:30 (#0)
  -> non_recursive_function_b at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:12:1: 12:30 (#0)

@@@ targets @@@

tests -(0)-> recursive_function at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:6:1: 6:32 (#0)
  (0) test

tests -(2)-> non_recursive_function_a at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:11:1: 11:30 (#0)
  (2) test

tests -(2)-> non_recursive_function_b at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:12:1: 12:30 (#0)
  (2) test

tests -(2)-> non_recursive_function_c at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:14:1: 14:30 (#0)
  (2) test

tests -(1)-> make_indirect_recursive_call at tests/ui/call_graph/avoid_counting_indirect_cycles.rs:19:1: 19:42 (#0)
  (1) test

targets: 5 total; 5 safe; 0 unsafe (0 tainted)
